spec/jmap/push.mdown

```diff
@@ -19,50 +19,125 @@ When something changes on the server, the server pushes a **StateChange** object
 
-Clients may register the push subscription with the JMAP server, which will then make a POST request to the associated push endpoint whenever an event occurs.
+Clients may create a push subscription on the JMAP server, which will then make a POST request to the associated push endpoint whenever an event occurs.
 
 The POST request MUST have a content type of `application/json` and contain the UTF-8 JSON encoded *StateChange* object as the body. The request MUST have a `TTL` header, and MAY have `Urgency` and/or `Topic` headers, as specified in section 5 of [@!RFC8030].
 
-If the response code is `503` (Service Unavailable), the JMAP server MAY try again later, but may also just drop the event. If the response code is `429` (Too Many Requests) the JMAP server SHOULD attempt to reduce the frequency of pushes to that URL. Any other `4xx` or `5xx` response code MUST be considered a **permanent failure** and the push subscription should be deregistered (not tried again even for future events unless explicitly re-registered by the client).
+If the response code is `503` (Service Unavailable), the JMAP server MAY try again later, but may also just drop the event. If the response code is `429` (Too Many Requests) the JMAP server SHOULD attempt to reduce the frequency of pushes to that URL. Any other `4xx` or `5xx` response code MUST be considered a **permanent failure** and the push subscription SHOULD be destroyed.
 
 The use of this push endpoint conforms with the use of a push endpoint by an Application Server as defined in [@!RFC8030]. A client MAY use the rest of [@!RFC8030] in combination with its own Push Server to form a complete end-to-end solution, or MAY rely on alternative mechanisms to ensure the delivery of the pushed data after it leaves the JMAP server.
 
-### PushSubscription/set
-
-Each session may only have a single push subscription registered. The push subscription is tied to the access token used to create it. Should the access token expire or be revoked, the push subscription MUST be removed by the JMAP server. The client MUST re-register the push subscription after reauthenticating to resume callbacks.
-
-To set the push subscription, make a call to *PushSubscription/set*. It takes the following argument:
-
-- **pushSubscription**: `PushSubscription|null`
-  The PushSubscription object representing the endpoint the JMAP server will POST events to. This will replace any previously set subscription. Set to `null` to remove any previously registered subscription.
+The push subscription is tied to the credentials used to authenticate the API request that created it. Should these credentials expire or be revoked, the push subscription MUST be destroyed by the JMAP server.
```

Comment:
Could we limit the requirement to just destroying the subscription on credential revocation, and not necessarily expiration?

This is the rosy future scenario I'm thinking of: you're an iOS developer, connecting to a Google server using Oauth. And the APNS of the future supports RFC 8030 semantics, and the JMAP server can push directly without an intermediate proxy server. Yet your Oauth access tokens continue to expire frequently after using your refresh token to get one, on the order of ~24 hours. 

If the PushSubscription expires when the access token expires, it might expire within 24h, or less. On iOS, it's a little difficult to guarantee background execution for an app. Nothing is airtight (users can disable background app refresh altogether), but a scenario I like to plan for is the occasional use of low power mode, which disables background app refresh for the day. In that case, you need to wait until a full device charge cycle happens until you can execute in the background, so 48h is a better lower bound to plan for than 24h.

Or in this case, were you thinking that the server would only expire the PushSubscription when the refresh token expires, not the access token? 

Thoughts?

(Maybe another avenue to consider is a final "SubscriptionWillExpire" push right before the subscription expires, so devices/services can do whatever renewal is needed).

Reply:
> Or in this case, were you thinking that the server would only expire the PushSubscription when the refresh token expires, not the access token?

Yes, I really mean the push subscription must be destroyed when the "session" expires, which is really the refresh token rather than the access token in your Oauth scenario. I'm not sure how best to write that here without going into Oauth specifics, which I don't really want to do.